Move files/ajax/scan.php to background job

LukasReschke opened the pull request:
The background job will now be executed in chunks of 500 users all 10 minutes.

@DeepDiver1975 @icewind1991 @PVince81  Please review - THX.

Fixes https://github.com/owncloud/core/issues/17906

LukasReschke committed:
Move files/ajax/scan.php to background job

The background job will now be executed in chunks of 500 users all 10 minutes.

LukasReschke commented on apps/files/lib/backgroundjob/scanfiles.php:
```diff
@@ -0,0 +1,100 @@
+	 */
+	protected function runScanner(IUser $user) {
+		$scanner = new Scanner($user->getUID(), $this->dbConnection);
+		$scanner->scan();
```
@icewind1991 @PVince81 Please advise whether we should use `\OC\Files\Utils\Scanner::backgroundScan` or `\OC\Files\Utils\Scanner::scan`

LukasReschke replied:
Especially as this will be run all 10 minutes for 500 users. :wink:

LukasReschke commented on apps/files/lib/backgroundjob/scanfiles.php:
```diff
@@ -0,0 +1,100 @@
+
+	/**
+	 * @param IUser $user
+	 * @throws \OC\ForbiddenException
```
It also seems that the scanner can throw a forbidden exception. Shall we be graceful here and then just skip the user and add the exception to the log?

LukasReschke replied:
Done

karlitschek commented:
looks good. 500 users every 10min sounds reasonable

nickvergessen commented on apps/files/appinfo/update.php:
```diff
@@ -96,6 +96,14 @@ function owncloud_reset_encrypted_flag(\OCP\IDBConnection $conn) {
+// Add cron job for scanning user storages
+$jobList = \OC::$server->getJobList();
+$job = 'OCA\Files\BackgroundJob\ScanFiles';
+if(!$jobList->has($job, '')) {
```
Just add it, the JobList takes care of deduplication:
https://github.com/owncloud/core/blob/master/lib/private/backgroundjob/joblist.php#L54-L55

LukasReschke replied:
Done. THX.

LukasReschke commented:
https://github.com/owncloud/core/pull/20545#discussion_r44986186 is still open

icewind1991 commented:
Add some tests for the iterating over users cases?

- Offset is at the end of the user list, job should reset offset and query again
- With >500 users 2 consecutive calls should get a different list of users
- With <500 users 2 consecutive calls should get the same list of users

icewind1991 commented on apps/files/lib/backgroundjob/scanfiles.php:
```diff
@@ -0,0 +1,110 @@
+	protected function runScanner(IUser $user) {
+		try {
+			$scanner = new Scanner($user->getUID(), $this->dbConnection);
+			$scanner->scan();
```
scan.php uses `$scanner->backgroundScan();` which is also what should be used here.

There different is that `backgroundScan` only checks for folders that are not fully scanned yet (marked with size -1) while `scan` goes trough the entire filesystem of the user.

Using `scan` in a backgroundjob in such a naive way will not work when you scale up, when external storages are used `scan` can easily take many minutes every time you call it, `backgroundScan` will only take time if there's work to be done

PVince81 commented:
Rebase needed

LukasReschke commented:
Rebased also added a new unit test.

> - With >500 users 2 consecutive calls should get a different list of users
> - With <500 users 2 consecutive calls should get the same list of users

@icewind1991 I'm not entirely sure how to do that here considering the mocking and DIs. So what I just added now is a test that it goes back to the first user once it has reached all.

MorrisJobke commented:
Tested and works :+1:
